juju/juju
# feat: charm domain state

The change set adds simple getter and setter methods, along with getting the charm metadata. Getting the charm metadata from the database is quite involved. We have to select from a lot of tables to populate the metadata. This is really wasteful if the code only wants the relations. It is also not time efficient to get all the metadata as row based type. If a charm metadata is fully populated it could result in thousands of rows to populate. Having spoken with others, it was then decided to not use row-based requests for getting metadata. Instead, we'll fallback to multiple selects.
The advantage of doing it this way is that we can reuse the individual methods to expose calls to get just relations or
storage.

There is a potential problem with arrays. Duplicates of values are possible in the charm metadata and we don't prevent that in any part of the library. Also, tables are sets of rows, not ordered rows based on insertion time. The solution was to keep a tracking index column that ensures that we preserve the insertion order and that we use that as the unique key.

Unfortunately, I ran into a sqlair bug[1], so had to change tack and ensure that we could get the right information out that we wanted. This meant that we stuck to the error conditions. If the charm doesn't exist we return NotFound, rather than the default case.

Actions, config, manifest, and LXD profiles will be added in a future PR, as this is already getting difficult to review in one patch.


 1. https://github.com/canonical/sqlair/issues/154



## Checklist




## QA steps

The schema should apply nicely.

Regression tests around bootstrap and deployment should be done.

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ juju deploy ubuntu
```

## Links

**Jira card:** [JUJU-6015](https://warthogs.atlassian.net/browse/JUJU-6015)



[JUJU-6015]: https://warthogs.atlassian.net/browse/JUJU-6015?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
